src/librustc_mir/monomorphize/polymorphize.rs

```diff
@@ -0,0 +1,283 @@
+    tcx: TyCtxt<'tcx>,
+    def_id: DefId,
+    body: ReadOnlyBodyAndCache<'a, 'tcx>,
+    used_parameters: &'a mut BitSet<u32>,
+}
+
+impl<'a, 'tcx> UsedGenericParametersVisitor<'a, 'tcx> {
+    fn visit_foldable<T: std::fmt::Debug + TypeFoldable<'tcx>>(&mut self, ty: T) {
+        debug!("visit_foldable: ty={:?}", ty);
+        // Exit early if this type doesn't use any type parameters.
+        if !ty.has_param_types() {
+            return;
+        }
+
+        // Use `TypeVisitor` to find any type parameters.
+        ty.visit_with(self);
+    }
+}
+
+impl<'a, 'tcx> Visitor<'tcx> for UsedGenericParametersVisitor<'a, 'tcx> {
+    fn visit_local_decl(&mut self, local: Local, local_decl: &LocalDecl<'tcx>) {
+        debug!("visit_local_decl: local_decl={:?}", local_decl);
+        let ty = local_decl
+            .ty
+            .builtin_deref(false)
+            .map(|type_and_mut| type_and_mut.ty)
+            .unwrap_or(local_decl.ty);
+        debug!("visit_local_decl: ty={:?}", ty);
+        // Skip visiting the closure that is currently being processed. This only happens
+        // because the first argument to the closure is a reference to itself and that
+        // will call `visit_substs`, resulting in each generic parameter captured being
+        // considered used by default.
+        match ty.kind {
+            ty::Closure(closure_def_id, _) if closure_def_id == self.def_id => {
+                debug!("visit_local_decl: skipping closure substs");
+                return;
+            }
+            _ => {}
+        }
+
+        self.super_local_decl(local, local_decl);
+    }
+
+    fn visit_const(&mut self, c: &&'tcx ty::Const<'tcx>, _: Location) {
+        debug!("visit_const: c={:?}", c);
+        // Necessary to detect `T::ASSOC_CONST`.
+        self.visit_foldable(*c);
+    }
+
+    fn visit_rvalue(&mut self, rvalue: &Rvalue<'tcx>, location: Location) {
+        debug!("visit_rvalue: rvalue={:?}", rvalue);
+        if let Rvalue::Cast(CastKind::Pointer(PointerCast::Unsize), ..) = *rvalue {
+            // If a closure performs dynamic dispatch then all of its generic parameters are
+            // used. This shouldn't be necessary but until there are changes to vtable
+            // infrastructure, it is.
+            self.used_parameters.insert_all();
+            return;
+        }
+
+        self.super_rvalue(rvalue, location);
+    }
+
+    fn visit_ty(&mut self, ty: Ty<'tcx>, _: TyContext) {
+        self.visit_foldable(ty);
+    }
+
+    fn visit_terminator_kind(&mut self, kind: &TerminatorKind<'tcx>, _: Location) {
+        if let TerminatorKind::Call { func, .. } = kind {
+            debug!("visit_terminator_kind: func={:?}", func);
+            if let ty::FnDef(_, substs) = func.ty(&self.body.local_decls, self.tcx).kind {
+                debug!("visit_terminator_kind: substs={:?}", substs);
+                self.visit_foldable(substs);
+            }
+        }
+    }
```

Comment:
This shouldn't be needed since you implemented `visit_const` correctly.

(Unless you want to make the analysis transitive, which would be useful but hard due to cycles, similar to inlining)

Reply:
I've fixed this in dd24b5b. I did want to make the analysis transitive, but I stopped when I figured out that cycles would be an issue.